File: backend/apps/api/routers/jenkins_jobs.py
from __future__ import annotations

from ninja import Router
from ninja.errors import HttpError

from apps.buildmeta.selectors.jenkins import list_executions_by_job_path, list_jenkins_job_summaries
from apps.buildmeta.services.jenkins import (
    JenkinsOperationError,
    get_jenkins_build_details,
    get_jenkins_client_config,
    get_jenkins_job_details,
    get_jenkins_job_status,
    trigger_jenkins_job,
)


router = Router(tags=["jenkins-jobs"])


@router.get("/")
def list_jenkins_jobs(request) -> list[dict]:
    return list_jenkins_job_summaries(ci_provider="jenkins")


@router.get("/{job_path}/status")
def get_jenkins_job_status_endpoint(request, job_path: str) -> dict:
    try:
        return get_jenkins_job_status(job_path)
    except JenkinsOperationError as exc:
        raise HttpError(400, exc.summary)


@router.get("/{job_path}/details")
def get_jenkins_job_details_endpoint(request, job_path: str) -> dict:
    try:
        return get_jenkins_job_details(job_path)
    except JenkinsOperationError as exc:
        raise HttpError(404, exc.summary)


@router.get("/{job_path}/executions")
def get_jenkins_executions(request, job_path: str) -> list[dict]:
    payload = list_executions_by_job_path(job_path)
    if payload is None:
        raise HttpError(404, f"Jenkins job '{job_path}' was not found.")
    return payload


@router.post("/{job_path}/trigger")
def trigger_jenkins_job_endpoint(request, job_path: str, parameters: dict | None = None) -> dict:
    try:
        return trigger_jenkins_job(job_path, parameters)
    except JenkinsOperationError as exc:
        raise HttpError(400, exc.summary)


@router.get("/{job_path}/builds/{build_number}")
def get_jenkins_build_endpoint(request, job_path: str, build_number: str) -> dict:
    try:
        return get_jenkins_build_details(job_path, build_number)
    except JenkinsOperationError as exc:
        raise HttpError(404, exc.summary)


@router.get("/system-status")
def get_jenkins_system_status(request) -> dict:
    from apps.buildmeta.services.jenkins import collect_jenkins_system_status
    try:
        collect_jenkins_system_status()
    except JenkinsOperationError as exc:
        raise HttpError(400, exc.summary)
    config = get_jenkins_client_config()
    return {
        "serverUrl": config.server_url,
        "connected": True,
    }

File: backend-rs/src/api/routes/jenkins_jobs.rs
use actix_web::{web, HttpResponse, Result};
use serde::Deserialize;

use crate::api::responses;
use crate::infrastructure::repositories::jenkins_repo;
use crate::AppState;

#[derive(Debug, Deserialize)]
pub struct TriggerRequest {
    #[serde(default)]
    pub parameters: serde_json::Value,
}

pub async fn list_jenkins_jobs(state: web::Data<AppState>) -> Result<HttpResponse> {
    let Some(pool) = state.pool.as_ref() else {
        return Ok(responses::service_unavailable_database());
    };
    match jenkins_repo::list_jenkins_job_summaries(pool).await {
        Ok(payload) => Ok(HttpResponse::Ok().json(payload)),
        Err(error) => Ok(responses::internal_error(error.to_string())),
    }
}

pub async fn get_jenkins_job_status_endpoint(
    state: web::Data<AppState>,
    path: web::Path<String>,
) -> Result<HttpResponse> {
    let job_path = path.into_inner();
    let Some(pool) = state.pool.as_ref() else {
        return Ok(responses::service_unavailable_database());
    };
    match jenkins_repo::get_jenkins_job_status(pool, &job_path).await {
        Ok(Some(payload)) => Ok(HttpResponse::Ok().json(payload)),
        Ok(None) => Ok(responses::not_found("Jenkins job was not found.")),
        Err(error) => Ok(responses::map_domain_error(&error)),
    }
}

pub async fn get_jenkins_job_details_endpoint(
    state: web::Data<AppState>,
    path: web::Path<String>,
) -> Result<HttpResponse> {
    let job_path = path.into_inner();
    let Some(pool) = state.pool.as_ref() else {
        return Ok(responses::service_unavailable_database());
    };
    match jenkins_repo::get_jenkins_job_details(pool, &job_path).await {
        Ok(Some(payload)) => Ok(HttpResponse::Ok().json(payload)),
        Ok(None) => Ok(responses::not_found("Jenkins job was not found.")),
        Err(error) => Ok(responses::map_domain_error(&error)),
    }
}

pub async fn get_jenkins_executions(
    state: web::Data<AppState>,
    path: web::Path<String>,
) -> Result<HttpResponse> {
    let job_path = path.into_inner();
    let Some(pool) = state.pool.as_ref() else {
        return Ok(responses::service_unavailable_database());
    };
    match jenkins_repo::list_executions_by_job_path(pool, &job_path).await {
        Ok(payload) => Ok(HttpResponse::Ok().json(payload)),
        Err(error) => Ok(responses::map_domain_error(&error)),
    }
}

pub async fn trigger_jenkins_job_endpoint(
    state: web::Data<AppState>,
    path: web::Path<String>,
    web::Json(payload): web::Json<TriggerRequest>,
) -> Result<HttpResponse> {
    let job_path = path.into_inner();
    let Some(pool) = state.pool.as_ref() else {
        return Ok(responses::service_unavailable_database());
    };
    match jenkins_repo::trigger_jenkins_job(pool, &job_path, &payload.parameters).await {
        Ok(Some(result)) => Ok(HttpResponse::Ok().json(result)),
        Ok(None) => Ok(responses::not_found("Jenkins job was not found.")),
        Err(error) => Ok(responses::map_domain_error(&error)),
    }
}

pub async fn configure_jenkins_job_endpoint(
    state: web::Data<AppState>,
    path: web::Path<String>,
) -> Result<HttpResponse> {
    let job_path = path.into_inner();
    let Some(pool) = state.pool.as_ref() else {
        return Ok(responses::service_unavailable_database());
    };
    match jenkins_repo::configure_jenkins_job(pool, &job_path).await {
        Ok(Some(result)) => Ok(HttpResponse::Ok().json(result)),
        Ok(None) => Ok(responses::not_found("Jenkins job was not found.")),
        Err(error) => Ok(responses::map_domain_error(&error)),
    }
}

pub async fn get_jenkins_build_endpoint(
    state: web::Data<AppState>,
    path: web::Path<(String, String)>,
) -> Result<HttpResponse> {
    let (job_path, build_number) = path.into_inner();
    let Some(pool) = state.pool.as_ref() else {
        return Ok(responses::service_unavailable_database());
    };
    match jenkins_repo::get_jenkins_build_details(pool, &job_path, &build_number).await {
        Ok(Some(payload)) => Ok(HttpResponse::Ok().json(payload)),
        Ok(None) => Ok(responses::not_found("Jenkins build was not found.")),
        Err(error) => Ok(responses::map_domain_error(&error)),
    }
}

pub async fn get_jenkins_system_status(state: web::Data<AppState>) -> Result<HttpResponse> {
    let Some(pool) = state.pool.as_ref() else {
        return Ok(responses::service_unavailable_database());
    };
    match jenkins_repo::get_jenkins_system_status(pool).await {
        Ok(payload) => Ok(HttpResponse::Ok().json(payload)),
        Err(error) => Ok(responses::map_domain_error(&error)),
    }
}

pub fn configure(cfg: &mut web::ServiceConfig) {
    cfg.service(web::resource("/jenkins-jobs").route(web::get().to(list_jenkins_jobs)))
        .service(web::resource("/jenkins-jobs/system-status").route(web::get().to(get_jenkins_system_status)))
        .service(web::resource("/jenkins-jobs/{job_path:.+}/builds/{build_number}").route(web::get().to(get_jenkins_build_endpoint)))
        .service(web::resource("/jenkins-jobs/{job_path:.+}/status").route(web::get().to(get_jenkins_job_status_endpoint)))
        .service(web::resource("/jenkins-jobs/{job_path:.+}/details").route(web::get().to(get_jenkins_job_details_endpoint)))
        .service(web::resource("/jenkins-jobs/{job_path:.+}/executions").route(web::get().to(get_jenkins_executions)))
        .service(web::resource("/jenkins-jobs/{job_path:.+}/trigger").route(web::post().to(trigger_jenkins_job_endpoint)))
        .service(web::resource("/jenkins-jobs/{job_path:.+}/configure").route(web::post().to(configure_jenkins_job_endpoint)));
}
